File: pi/lib/herdr_workflow/tests/fakes.py
"""Test doubles for the Herdr/Git/Clock/TraceExporter seams, plus a tmp-repo helper.

ponytail: real git in a tmp dir beats mocking diff output; upgrade to a pure
fake only if git-in-CI proves flaky.
"""
import subprocess
from datetime import datetime, timedelta, timezone
from pathlib import Path

from herdr_workflow import effects

DEFAULT_CONFIG = {
    "models": {
        "worker_default": "test/worker",
        "verifier": "test/verifier",
        "verifier_fallback": "test/verifier-fallback",
        "archive": "test/archive",
        "git": "test/git",
        "planner": "test/planner",
        "triage": "test/triage",
        "recovery": "test/recovery",
    },
    "thinking": {
        "worker_default": "high",
        "verifier": "high",
        "verifier_lite": "medium",
        "planner": "high",
        "triage": "high",
        "recovery": "high",
        "archive": "high",
    },
    "workflow": {
        "max_verification_rounds": 6,
        "verification_timeout_seconds": 600,
        "remote": "origin",
        "branch_prefix": "feature/",
        "base_branch": "origin/HEAD",
    },
    "projects": {"root": "~/development", "max_depth": 3},
    "plugins": {},
}


class FakeHerdr:
    """Records every call(*args); returns scripted or generated responses.

    Agent state is tracked by pane_id (the source of truth), matching the real
    herdr API where `wait agent-status`, `pane run`, and `pane send-keys` all
    target a pane_id and `agent get` has no sequence counter (no
    `state_change_seq` field) — only `agent_status`. `agent get <name>` resolves
    name -> pane_id via the `agent rename` mapping, exactly as production code
    relies on.
    """

    def __init__(self):
        self.calls = []
        self._pane_to_agent = {}   # pane_id -> name, set by `agent rename` (or register_pane)
        self._agent_to_pane = {}   # reverse of the above
        self._pane_status = {}     # pane_id -> agent_status; absent until first touched
        self._tab_seq = 0
        self._pane_seq = 0
        self.handlers = []      # [(predicate(args) -> bool, handler(args) -> dict)]
        self.after_hooks = []   # [(predicate(args) -> bool, effect(args) -> None)]
        # Default: `pane run` settles the target pane into the status a real
        # submission would reach — "idle" for a pi cold-boot (first `pane run` on a
        # pane), "working" for a prompt on an already-launched agent (any later
        # `pane run`). Tests exercising the verification/nudge/failure paths set
        # this False and script transitions explicitly via `.after(...)`.
        self.auto_advance_on_submit = True

    def on(self, predicate, handler):
        self.handlers.append((predicate, handler))

    def after(self, predicate, effect):
        self.after_hooks.append((predicate, effect))

    def register_pane(self, pane_id, name):
        """Associate a pane with an agent name, as `agent rename` would."""
        self._pane_to_agent[pane_id] = name
        self._agent_to_pane[name] = pane_id

    def set_agent(self, name, pane_id=None, agent_status=None, **_ignored):
        """Test convenience: register/update an agent's pane and/or status directly."""
        if pane_id:
            self.register_pane(pane_id, name)
        target_pane = pane_id or self._agent_to_pane.get(name)
        if agent_status is not None and target_pane:
            self._pane_status[target_pane] = agent_status

    def set_status(self, pane_id_or_name, status):
        pane_id = self._agent_to_pane.get(pane_id_or_name, pane_id_or_name)
        self._pane_status[pane_id] = status

    def call(self, *args):
        self.calls.append(args)
        result = None
        matched = False
        for predicate, handler in self.handlers:
            if predicate(args):
                result = handler(args)
                matched = True
                break
        if not matched:
            result = self._default(args)
        for predicate, effect in self.after_hooks:
            if predicate(args):
                effect(args)
        return result

    def _default(self, args):
        if args[:2] == ("tab", "create"):
            self._tab_seq += 1
            self._pane_seq += 1
            tab_id, pane_id = f"tab-{self._tab_seq}", f"pane-{self._pane_seq}"
            # Real `herdr tab create` result has no top-level tab_id: it's nested in
            # root_pane (and in "tab"). Both launch_role and create_tab() read it
            # from root_pane["tab_id"].
            return {"root_pane": {"pane_id": pane_id, "tab_id": tab_id}, "tab": {"tab_id": tab_id}}
        if args[:2] == ("agent", "start"):
            # Real `agent start` launches inside the caller-supplied `--pane`; it
            # never fabricates a new pane. Registering anything else here would
            # desync `state["panes"][role]` from what `agent get` resolves.
            name = args[2]
            pane_id = args[args.index("--pane") + 1]
            self.register_pane(pane_id, name)
            self._pane_status[pane_id] = "idle"
            return {"agent": {"pane_id": pane_id, "name": name}}
        if args[:2] == ("agent", "prompt"):
            name = args[2]
            pane_id = self._agent_to_pane.get(name)
            if pane_id is None:
                raise SystemExit(f"agent not found: {name}")
            if self.auto_advance_on_submit:
                self._pane_status[pane_id] = "working"
            return {"agent": {"pane_id": pane_id, "agent_status": self._pane_status.get(pane_id, "working")}}
        if args[:2] == ("agent", "rename"):
            self.register_pane(args[2], args[3])
            return {}
        if args[:2] == ("pane", "process-info"):
            return {"process_info": {"foreground_processes": [{"name": "zsh"}]}}
        if args[:2] == ("pane", "read"):
            return {"read": {"text": "❯ "}}
        if args[:2] == ("pane", "run"):
            pane_id = args[2]
            if self.auto_advance_on_submit:
                # First `pane run` on a pane is the pi launch (settles to idle once
                # booted); any later one is a prompt on an already-running agent
                # (settles to working).
                self._pane_status[pane_id] = "working" if pane_id in self._pane_status else "idle"
            return {}
        if args[:2] in {("pane", "close"), ("pane", "send-keys"), ("notification", "show")}:
            return {}
        if args[:2] == ("agent", "get"):
            name = args[2]
            pane_id = self._agent_to_pane.get(name)
            if pane_id is None:
                raise SystemExit(f"agent not found: {name}")
            return {"agent": {"agent_status": self._pane_status.get(pane_id, "idle"), "pane_id": pane_id}}
        if args[:2] == ("wait", "agent-status"):
            # herdr wait agent-status <pane_id> --status <status> --timeout <ms>
            pane_id = args[2]
            wanted = args[args.index("--status") + 1]
            if self._pane_status.get(pane_id, "idle") == wanted:
                return {}
            raise SystemExit("timed out waiting for agent status change")
        if args[0] == "wait":
            return {}
        return {}


class FakeGit:
    """Backed by a real temp git repo — cheapest correct option for diff/numstat parsing."""

    def run(self, *args, cwd):
        result = subprocess.run(["git", *args], cwd=cwd, text=True, capture_output=True, check=False)
        if result.returncode:
            raise SystemExit(f"git {' '.join(args)}: {(result.stderr or result.stdout or 'command failed').strip()}")
        return result.stdout.strip()


class FakeClock:
    """Fixed/monotone: `sleep` advances virtual time so timeout logic is deterministic and fast."""

    def __init__(self, start=1_700_000_000.0):
        self._now = datetime(2024, 1, 1, tzinfo=timezone.utc)
        self._monotonic = start
        self._time = start

    def now(self):
        return self._now

    def monotonic(self):
        return self._monotonic

    def time(self):
        return self._time

    def time_ns(self):
        return int(self._time * 1_000_000_000)

    def sleep(self, seconds):
        self._monotonic += seconds
        self._time += seconds
        self._now += timedelta(seconds=seconds)

    def advance(self, seconds):
        self.sleep(seconds)


class NoopExporter:
    def export(self, record):
        pass


def make_context(config=None, herdr=None, git=None, clock=None, exporter=None):
    return effects.Context(
        config=config if config is not None else DEFAULT_CONFIG,
        herdr=herdr if herdr is not None else FakeHerdr(),
        git=git if git is not None else FakeGit(),
        clock=clock if clock is not None else FakeClock(),
        exporter=exporter if exporter is not None else NoopExporter(),
    )


def init_repo(path):
    """Init a git repo with an OpenSpec project and one committed base file."""
    path = Path(path)
    path.mkdir(parents=True, exist_ok=True)
    run = lambda *args: subprocess.run(["git", *args], cwd=path, check=True, capture_output=True, text=True)
    run("init", "-q", "-b", "main")
    run("config", "user.email", "test@example.com")
    run("config", "user.name", "Test")
    (path / "openspec").mkdir(exist_ok=True)
    (path / "openspec" / "config.yaml").write_text("name: test\n")
    (path / "README.md").write_text("# test\n")
    run("add", "-A")
    run("commit", "-q", "-m", "base")
    (path / ".git" / "info" / "exclude").write_text("\n.herdr-workflow/\n")
    return path

File: agentic-coding/test/fakes.ts
// Test doubles for the Herdr/Git/Clock/TraceExporter seams, plus a tmp-repo helper.
//
// ponytail: real git in a tmp dir beats mocking diff output; upgrade to a pure
// fake only if git-in-CI proves flaky.
import { execFileSync } from 'node:child_process';
import fs from 'node:fs';
import path from 'node:path';
import type { Context } from '../src/workflow/effects.ts';

export const DEFAULT_CONFIG = {
  models: {
    worker_default: 'test/worker',
    verifier: 'test/verifier',
    usability_verifier: 'test/usability',
    verifier_fallback: 'test/verifier-fallback',
    archive: 'test/archive',
    git: 'test/git',
    planner: 'test/planner',
    triage: 'test/triage',
    recovery: 'test/recovery',
  },
  thinking: {
    worker_default: 'high',
    verifier: 'high',
    verifier_lite: 'medium',
    planner: 'high',
    triage: 'high',
    recovery: 'high',
    archive: 'high',
  },
  workflow: {
    max_verification_rounds: 6,
    remote: 'origin',
    branch_prefix: 'feature/',
    base_branch: 'origin/HEAD',
    worktree_directory: '~/.herdr/worktrees',
  },
  projects: { root: '~/development', max_depth: 3 },
  telemetry: { capture_content: false },
  ui: { theme: 'catppuccin', selection_height: 10 },
  plugins: {},
};

type Predicate = (args: string[]) => boolean;
type Handler = (args: string[]) => any;

/**
 * Records every call(...args); returns scripted or generated responses.
 *
 * Agent state is tracked by pane_id (the source of truth), matching the real
 * herdr API where `wait agent-status`, `pane run`, and `pane send-keys` all
 * target a pane_id and `agent get` has no sequence counter — only `agent_status`.
 * `agent get <name>` resolves name -> pane_id via the `agent rename` mapping,
 * exactly as production code relies on.
 */
export class FakeHerdr {
  calls: string[][] = [];
  private paneToAgent = new Map<string, string>();
  private agentToPane = new Map<string, string>();
  private paneToTab = new Map<string, string>();
  private paneStatus = new Map<string, string>();
  private tabSeq = 0;
  private paneSeq = 0;
  private handlers: Array<[Predicate, Handler]> = [];
  private afterHooks: Array<[Predicate, (args: string[]) => void]> = [];
  // Default: `pane run` settles the target pane into the status a real submission
  // would reach. Tests exercising verification/failure paths set this false and
  // script transitions explicitly via `.after(...)`.
  autoAdvanceOnSubmit = true;

  on(predicate: Predicate, handler: Handler): void {
    this.handlers.push([predicate, handler]);
  }

  after(predicate: Predicate, effect: (args: string[]) => void): void {
    this.afterHooks.push([predicate, effect]);
  }

  registerPane(paneId: string, name: string, tabId?: string): void {
    this.paneToAgent.set(paneId, name);
    this.agentToPane.set(name, paneId);
    if (tabId) this.paneToTab.set(paneId, tabId);
  }

  setAgent(name: string, options: { paneId?: string; agentStatus?: string } = {}): void {
    if (options.paneId) this.registerPane(options.paneId, name);
    const targetPane = options.paneId ?? this.agentToPane.get(name);
    if (options.agentStatus !== undefined && targetPane) this.paneStatus.set(targetPane, options.agentStatus);
  }

  setStatus(paneIdOrName: string, status: string): void {
    const paneId = this.agentToPane.get(paneIdOrName) ?? paneIdOrName;
    this.paneStatus.set(paneId, status);
  }

  get paneToTabMap(): Map<string, string> {
    return this.paneToTab;
  }

  call(...args: string[]): any {
    this.calls.push(args);
    let result: any;
    let matched = false;
    for (const [predicate, handler] of this.handlers) {
      if (predicate(args)) {
        result = handler(args);
        matched = true;
        break;
      }
    }
    if (!matched) result = this.default(args);
    for (const [predicate, effect] of this.afterHooks) {
      if (predicate(args)) effect(args);
    }
    return result;
  }

  private default(args: string[]): any {
    if (args[0] === 'tab' && args[1] === 'create') {
      this.tabSeq += 1;
      this.paneSeq += 1;
      const tabId = `tab-${this.tabSeq}`;
      const paneId = `pane-${this.paneSeq}`;
      this.paneToTab.set(paneId, tabId);
      return { root_pane: { pane_id: paneId, tab_id: tabId }, tab: { tab_id: tabId } };
    }
    if (args[0] === 'pane' && args[1] === 'split') {
      this.paneSeq += 1;
      const paneId = `pane-${this.paneSeq}`;
      this.paneToTab.set(paneId, this.paneToTab.get(args[2])!);
      return { pane: { pane_id: paneId, tab_id: this.paneToTab.get(paneId) } };
    }
    if (args[0] === 'agent' && args[1] === 'start') {
      const name = args[2];
      const paneId = args[args.indexOf('--pane') + 1];
      const tabId = this.paneToTab.get(paneId);
      this.registerPane(paneId, name, tabId);
      this.paneStatus.set(paneId, 'idle');
      return { agent: { pane_id: paneId, tab_id: tabId, name, agent_status: 'idle' } };
    }
    if (args[0] === 'agent' && args[1] === 'prompt') {
      const target = args[2];
      const paneId = this.paneToAgent.has(target) || this.paneStatus.has(target) ? target : this.agentToPane.get(target);
      if (paneId == null) throw new Error(`agent not found: ${target}`);
      if (this.autoAdvanceOnSubmit) this.paneStatus.set(paneId, 'working');
      return { agent: { pane_id: paneId, agent_status: this.paneStatus.get(paneId) ?? 'working' } };
    }
    if (args[0] === 'agent' && args[1] === 'rename') {
      this.registerPane(args[2], args[3]);
      return {};
    }
    if (args[0] === 'pane' && args[1] === 'get') {
      return { pane: { pane_id: args[2], tab_id: this.paneToTab.get(args[2]) } };
    }
    if (args[0] === 'pane' && args[1] === 'process-info') {
      return { process_info: { foreground_processes: [{ name: 'zsh' }] } };
    }
    if (args[0] === 'pane' && args[1] === 'read') {
      return { read: { text: '❯ ' } };
    }
    if (args[0] === 'pane' && args[1] === 'run') {
      const paneId = args[2];
      if (this.autoAdvanceOnSubmit) {
        // First `pane run` on a pane is the pi launch (settles to idle once
        // booted); any later one is a prompt on an already-running agent
        // (settles to working).
        this.paneStatus.set(paneId, this.paneStatus.has(paneId) ? 'working' : 'idle');
      }
      return {};
    }
    if ((args[0] === 'pane' && (args[1] === 'close' || args[1] === 'send-keys')) || (args[0] === 'notification' && args[1] === 'show')) {
      return {};
    }
    if (args[0] === 'agent' && args[1] === 'get') {
      const target = args[2];
      const paneId = this.paneToAgent.has(target) || this.paneStatus.has(target) ? target : this.agentToPane.get(target);
      if (paneId == null) throw new Error(`agent not found: ${target}`);
      return { agent: { agent_status: this.paneStatus.get(paneId) ?? 'idle', pane_id: paneId, tab_id: this.paneToTab.get(paneId) } };
    }
    if (args[0] === 'wait' && args[1] === 'agent-status') {
      const paneId = args[2];
      const wanted = args[args.indexOf('--status') + 1];
      if ((this.paneStatus.get(paneId) ?? 'idle') === wanted) return {};
      throw new Error('timed out waiting for agent status change');
    }
    if (args[0] === 'wait') return {};
    return {};
  }
}

/** Backed by a real temp git repo — cheapest correct option for diff/numstat parsing. */
export class FakeGit {
  run(args: string[], cwd: string): string {
    try {
      return execFileSync('git', args, { cwd, encoding: 'utf8' }).trim();
    } catch (error: any) {
      const detail = (error.stderr?.toString() || error.stdout?.toString() || 'command failed').trim();
      throw new Error(`git ${args.join(' ')}: ${detail}`);
    }
  }
}

/** Fixed/monotone: `sleep` advances virtual time so timeout logic is deterministic and fast. */
export class FakeClock {
  private currentNow: Date;
  private monotonicSeconds: number;
  private timeSeconds: number;

  constructor(start = 1_700_000_000.0) {
    this.currentNow = new Date('2024-01-01T00:00:00Z');
    this.monotonicSeconds = start;
    this.timeSeconds = start;
  }

  now(): Date {
    return this.currentNow;
  }

  monotonic(): number {
    return this.monotonicSeconds;
  }

  time(): number {
    return this.timeSeconds;
  }

  timeNs(): bigint {
    return BigInt(Math.round(this.timeSeconds * 1_000_000_000));
  }

  async sleep(seconds: number): Promise<void> {
    this.advance(seconds);
  }

  advance(seconds: number): void {
    this.monotonicSeconds += seconds;
    this.timeSeconds += seconds;
    this.currentNow = new Date(this.currentNow.getTime() + seconds * 1000);
  }
}

export class NoopExporter {
  export(_record: unknown): void {
    /* no-op */
  }
}

export function makeContext(overrides: Partial<Context> = {}): Context {
  return {
    config: overrides.config ?? structuredClone(DEFAULT_CONFIG),
    herdr: overrides.herdr ?? new FakeHerdr(),
    git: overrides.git ?? new FakeGit(),
    clock: overrides.clock ?? new FakeClock(),
    exporter: overrides.exporter ?? new NoopExporter(),
  };
}

/** Init a git repo with an OpenSpec project and one committed base file. */
export function initRepo(dir: string): string {
  fs.mkdirSync(dir, { recursive: true });
  const run = (...args: string[]) => execFileSync('git', args, { cwd: dir, encoding: 'utf8' });
  run('init', '-q', '-b', 'main');
  run('config', 'user.email', 'test@example.com');
  run('config', 'user.name', 'Test');
  fs.mkdirSync(path.join(dir, 'openspec'), { recursive: true });
  fs.writeFileSync(path.join(dir, 'openspec', 'config.yaml'), 'name: test\n');
  fs.writeFileSync(path.join(dir, 'README.md'), '# test\n');
  run('add', '-A');
  run('commit', '-q', '-m', 'base');
  fs.mkdirSync(path.join(dir, '.git', 'info'), { recursive: true });
  fs.writeFileSync(path.join(dir, '.git', 'info', 'exclude'), '\n.herdr-workflow/\n');
  return dir;
}
